server/aesdsocket.c

```diff
@@ -0,0 +1,231 @@
+#include <fcntl.h>
+#include <syslog.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <sys/stat.h>
+
+
+#define CUSTOM_PORT "9000"
+#define MAX_CUSTOM_BUFFER 1024
+#define CUSTOM_LOG_FILE "/var/tmp/mysocketlog"
+
+int custom_socket_fd = -1;
+
+/* Function prototypes */
+void cleanup_resources();
+void signal_handler(int signo);
+void setup_logging();
+int create_socket();
+void set_socket_options(int sockfd);
+void bind_socket(int sockfd);
+void listen_for_connections(int sockfd);
+void handle_client(int client_fd);
+void accept_clients(int sockfd);
+void daemonize();
+
+// Cleans up resources on exit
+void cleanup_resources() {
+    syslog(LOG_INFO, "Cleaning up resources");
+    closelog();
+    if (custom_socket_fd != -1) {
+        close(custom_socket_fd);
+        custom_socket_fd = -1;
+    }
+    remove(CUSTOM_LOG_FILE);
+}
+
+// Signal handler for SIGINT and SIGTERM
+void signal_handler(int signo) {
+    syslog(LOG_USER, "Signal %d caught, exiting", signo);
+    cleanup_resources();
+    exit(EXIT_SUCCESS);
+}
+
+// Sets up logging using syslog
+void setup_logging() {
+    openlog("AESDSocket", LOG_PID | LOG_NDELAY, LOG_USER);
+}
+
+// Creates a socket and handles errors if any
+int create_socket() {
+    int sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd == -1) {
+        syslog(LOG_ERR, "Socket creation error: %m");
+        cleanup_resources();
+        exit(EXIT_FAILURE);
+    }
+    return sockfd;
+}
+
+// Sets socket options for reuse address and handles errors if any
+void set_socket_options(int sockfd) {
+    int optval = 1;
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) != 0) {
+        syslog(LOG_ERR, "Socket options setting error: %m");
+        cleanup_resources();
+        exit(EXIT_FAILURE);
+    }
+}
+
+// Binds the socket to a port and handles errors if any
+void bind_socket(int sockfd) {
+    struct sockaddr_in server_addr;
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_port = htons(atoi(CUSTOM_PORT));
+    server_addr.sin_addr.s_addr = INADDR_ANY;
+
+    if (bind(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) != 0) {
+        syslog(LOG_ERR, "Binding error: %m");
+        cleanup_resources();
+        exit(EXIT_FAILURE);
+    }
+}
+
+// Listens for incoming connections and handles errors if any
+void listen_for_connections(int sockfd) {
+    if (listen(sockfd, 5) != 0) {
+        syslog(LOG_ERR, "Listening error: %m");
+        cleanup_resources();
+        exit(EXIT_FAILURE);
+    }
+}
+
+// Handles client connections
+void handle_client(int client_fd) {
+    FILE *fp = fopen(CUSTOM_LOG_FILE, "a+");
+    char buffer[MAX_CUSTOM_BUFFER];
+
+    while (1) {
+        ssize_t bytes_recv = recv(client_fd, buffer, sizeof(buffer), 0);
+        if (bytes_recv <= 0) {
+            break;
+        }
+
+        fwrite(buffer, bytes_recv, 1, fp);
```

Comment:
@Suhas-Reddy-S I'm not sure the write before packet completes will work with future assignments.